### ISSUE TITLE
Documentation and ipynb examples are out of date

### ISSUE THREAD

**Issue by mivade** [aewallin/allantools#29]
I ran into problems while trying to run few-months-old code using allantools. It seems the latest release is backwards incompatible, but the documentation makes no mention of this. Furthermore, the Jupyter notebook examples are not updated to reflect this.

**Comment by aewallin** [comment]
Current documentation should be up-to-date with the current code and API - if it isn't we need a more precise bug report of where the documentation is wrong.

The API was changed abruptly and there is only a brief mention of this in the changelog at the top of allantools.py.

I don't use ipython or Jupyter so would need help with fixing the Jupyter examples.
The pure python examples should now work.

**Comment by mivade** [comment]
The [General usage](https://allantools.readthedocs.org/en/latest/api.html#general-usage) section of the API docs say that the inputs take `phase`, `rate`, and `taus` parameters, but it seems like in the latest release, all arguments are now keyword arguments and either `phase` or `frequency`. Since the docstrings of the functions contain a lot more information, it would also be nice to use [http://www.sphinx-doc.org/en/stable/ext/autodoc.html](autodoc) to include those in the documentation.

Regarding the notebook examples: the main advantage to those examples is that Github can render them directly so that the plots can be seen without having to run anything. Since they are otherwise identical to the script examples, I would recommend only having one or the other to avoid having to maintain both.

**Comment by fmeynadier** [comment]
Concerning the new style for function arguments : I like the fact that all arguments are named, but I find it a bit unusual to have two mutually exclusive inputs, with one being silently ignored if someone erroneously provides both.

A straightforward way to enforce the fact that one and only one input data is given would be to replace "phase" and "frequency" arguments by "data" and "type", the latter having only "phase" and "freq" as legal values (and possibly default to either one).

Such a change would mean 
- breaking the API (again, but impact to users would be limited if we do this shortly after previous break)
- a lot of refactoring, which I would be willing to do myself if Anders is OK with this idea.

**Comment by aewallin** [comment]
I don't know what the best solution for input parameters is. How is this solved in other projects? (scipy, pandas, or something similar)

In addition to simple arrays of phase and frequency maybe it would be good to support time-stamped data such as [ (t0,f0), (t1,f1), ... ]  where tn is a time-stamp and fn is a data-value. I think Stable32 has various strategies for filling missing data into gaps - this could be an enhancement at some point.

**Comment by mivade** [comment]
I think @fmeynadier's suggestion of having a single `data` argument is the most sensible for an immediate fix, provided that `type` would be given a default. This is fairly similar to how pandas typically does things where (often sensible) defaults are given, but a simple keyword argument switch will change the default behavior.

I also like the idea of a list of tuples as the data, especially if that is an optional way to input data. This shouldn't be too hard to implement if there is one required `data` argument since it is easy enough to check that the input is a list of tuples instead of a list of scalars.

So to summarize, function signatures could look something like this:

```python
allantools.adev(data, type="frequency", rate=1.0, taus=None)
```

**Comment by fmeynadier** [comment]
This way, time-stamped data could even have its own "type"...
However it is not clear to me how to keep numpy.array() efficiency with  [ (t0,f0), (t1,f1), ... ] data without casting it into arrays first, or even handle data whose samples are not aligned with a regular grid, with the current algorithms.

Anyway, what I propose is that I try to re-write the interface with the above-mentioned signature and see how it works.

Note that even naming the arguments is not mandatory when calling the function (unless we choose to enforce it, see for example [this "Effective Python" excerpt](http://www.informit.com/articles/article.aspx?p=2314818), item 21.

**Comment by aewallin** [comment]
I don't have a strong preference for the current function signature, so if many people think the suggestion above by @mivade is better and the way to go - then let's do that. In this case it is better to release a new version soon (say within April), so the current 2016.3 API is not out there very long. 
@fmeynadier: can you start a feature-branch with this work?

**Comment by fmeynadier** [comment]
Yes, I think I can do that quite quickly (~few days)

**Comment by fmeynadier** [comment]
This has been a lot quicker than expected, thanks to all your previous work ! (function calls in the test suite where already centralized).

See commit for a list of decisions I made w/r to the new signature : of course if you disagree with any of it, I can change that and propose a new commit.

**Comment by aewallin** [comment]
Thanks! this is merged now.
Before the next release I would like to see:

- Issue #35 solved
- improved test-coverage
- ipython notebooks either updated, or removed if nobody volunteers

**Comment by aewallin** [comment]
Thanks! closing this issue now.
Are there any other issues that need solving before the next release? (I want to release it soon, so that the old API is not out there too long)

**Comment by mivade** [comment]
Thanks!
